### PR TITLE
Data storage JavaScript SDK

### PR DESCRIPTION
**Pull request by ishachopra0401**
## Description of the change

Added new topic for JavaScript data storage by cookies or local storage

## Type of change
- [ ] Bug fix (non-breaking change that fixes an issue)
- [ ] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Related issues

> Fix [#1]() 

### Development

- [ ] Lint rules pass locally
- [ ] The code changed/added as part of this pull request has been covered with tests
- [ ] All tests related to the changed code pass in development

### Code review 

- [ ]  This pull request has a descriptive title and information useful to a reviewer. There may be a screenshot or screencast attached
- [ ] "Ready for review" label attached to the PR and reviewers mentioned in a comment
- [ ] Changes have been reviewed by at least one other engineer
- [ ] Issue from task tracker has a link to this pull request

### CODE REVIEW

File: stream-sources/rudderstack-sdk-integration-guides/rudderstack-javascript-sdk/data-storage-cookies.mdx [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+```
+
+## Cookies
+
+The list of cookies used by JavaScript SDK to store persistent user data is mentioned below:
+
+| Name             | Description       | Example    | Clearing Mechanism using SDK |
+| :-------------------- | :---------------------------------- | :-------------------------------- |:---------------------------------- |
+|  `rl_user_id`         |Stores the user ID set via the `identify` API method. All the subsequent event payloads will contain this data unless overridden. | `4578`, `USER_001` | `rudderanalytics.reset()`  |
+|  `rl_trait`       | Stores the user traits object set via the `identify` API method. All the subsequent event payloads will contain this data unless overridden. |  <pre>{<br>  email: "alex@example.com",<br>  accountType: "pro",<br>  country: "US",<br>  someObj: {<br>    key1: val1,<br>    key2: val2,<br>  }<br> }</pre>         |`rudderanalytics.reset()` |
```
Comment (saikumarrs) [inline stream-sources/rudderstack-sdk-integration-guides/rudderstack-javascript-sdk/data-storage-cookies.mdx:26]:
```suggestion
|  `rl_trait`       | Stores the user traits object set via the `identify` API method. All the subsequent event payloads will contain this data unless overridden. |  <pre>{<br>  email: "alex@example.com",<br>  accountType: "pro",<br>  country: "US",<br>  someObj: {<br>    key1: val1,<br>    key2: val2,<br>  }<br>}</pre>         |`rudderanalytics.reset()` |
```

File: stream-sources/rudderstack-sdk-integration-guides/rudderstack-javascript-sdk/data-storage-cookies.mdx [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+|  `rl_anonymous_id`          | Stores the anonymous ID. By default, it would be the auto-generated unique ID by SDK for each visitor unless overridden via `setAnonymousId` API method. All the subsequent event payloads will contain this data unless overridden. | `5bfe258f-bd2f-49cf-bddd-8b844f74ab4b`, `customAnonId`  | `rudderanalytics.reset(true)` |
```
Comment (saikumarrs) [inline stream-sources/rudderstack-sdk-integration-guides/rudderstack-javascript-sdk/data-storage-cookies.mdx:27]: For these 3 entries, update the last statement as below:

"All the subsequent event payloads will contain this data unless cleared from the storage."

File: stream-sources/rudderstack-sdk-integration-guides/rudderstack-javascript-sdk/data-storage-cookies.mdx [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+|  `rl_group_id`          | Stores the user group ID set via the `group` API method.        | `GRP_3`, `98` | `rudderanalytics.reset()` |
+|  `rl_group_trait`    | Stores the user group traits object set via the `group` API method.   |    <pre>{<br>  location: "New Orleans",<br>  nationality: "US",<br>  someObj: {<br>    key1: val1,<br>    key2: val2,<br>  }<br> }</pre>                      |  `rudderanalytics.reset()` |
```
Comment (saikumarrs) [inline stream-sources/rudderstack-sdk-integration-guides/rudderstack-javascript-sdk/data-storage-cookies.mdx:29]: For these 2 items, add the following statement to the description:

"All the subsequent group event payloads will contain this data unless cleared from the storage."

File: stream-sources/rudderstack-sdk-integration-guides/rudderstack-javascript-sdk/data-storage-cookies.mdx [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+
+
+## Local storage cookies
+
+The local storage cookies are used by the JavaScript SDK to keep track of the events being sent to the RudderStack backend. These can be listed as below:
+
+| Name             | Description       | Example    |
+| :------------- |:---------- | :----------- | 
+|  `ack`     | Timer for other browser tabs to claim control of the retry queue     | `1639734070124` |
+|  `reclaimStart`/`reclaimEnd`      |Determines if a tab takes over the queue from another tab  | `2dc2aee6-2836-4273-be69-79c90c04ddec`  |
```
Comment (saikumarrs) [inline stream-sources/rudderstack-sdk-integration-guides/rudderstack-javascript-sdk/data-storage-cookies.mdx:46]: These variables together contribute to the purpose.
```suggestion
|  `reclaimStart` and `reclaimEnd`      |Determines if a tab takes over the queue from another tab  | `2dc2aee6-2836-4273-be69-79c90c04ddec`  |
```

File: stream-sources/rudderstack-sdk-integration-guides/rudderstack-javascript-sdk/data-storage-cookies.mdx [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+|  `inProgress`    | Keeps track of the events in progress | <pre>{<br> "d89d7fb5-945e-4378-bda5-492e4b596fb4": {<br>  "item": {<br>   "url": "https://rudderstack-dataplane.rudderstack.com/v1/track",<br>    "headers": {<br>    "Content-Type": "application/json",<br>    ...<br> },<br> "message": { <br>   ...<br> }<br>  "attemptNumber": 1,<br>  "time": 1639734792773,<br>  "id": "a4d89d7f-b594-4eb3-b8bd-a5492e4b596f"<br> }<br> }<br>}<br></pre> |
+|  `queue`    | Keeps track of the events that are in queue to be processed   | <pre>[<br>  {<br>   "item": {<br>    "url": "https://rudderstack-dataplane.rudderstack.com/v1/track",<br>     "headers": {<br>     "Content-Type": "application/json",<br>    ...<br> },<br> "message": { <br>   ...<br> }<br>  "attemptNumber": 0,<br>  "time": 1639734792773,<br>  "id": "a4d89d7f-b594-4eb3-b8bd-a5492e4b596f"<br> }<br>]<br></pre> |
```
Comment (saikumarrs) [inline stream-sources/rudderstack-sdk-integration-guides/rudderstack-javascript-sdk/data-storage-cookies.mdx:48]: The examples are ill-formatted.

**Review by saikumarrs — CHANGES_REQUESTED** [review]
Added comments for improvements.

**Review by saikumarrs — APPROVED** [review]

File: stream-sources/rudderstack-sdk-integration-guides/rudderstack-javascript-sdk/data-storage-cookies.mdx [hunk 6]
```diff
@@ -0,0 +1,63 @@
+---
+title: "Data Storage in Cookies"
+description: >-
+  Detailed technical documentation on persistent user data stored in cookies or local storage by the RudderStack’s JavaScript SDK.
+---
+
+# Data Storage in Cookies
+
+RudderStack's JavaScript SDK stores persistent user data in cookies by default. In case the cookies are not supported, local storage is considered for the same.
```
Comment (ameypv-rudder) [inline stream-sources/rudderstack-sdk-integration-guides/rudderstack-javascript-sdk/data-storage-cookies.mdx:9]:
```suggestion
By default, the RudderStack JavaScript SDK stores persistent user data in the cookies. If the cookies are not supported, local storage is used instead. 
```

File: stream-sources/rudderstack-sdk-integration-guides/rudderstack-javascript-sdk/data-storage-cookies.mdx [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+title: "Data Storage in Cookies"
+description: >-
+  Detailed technical documentation on persistent user data stored in cookies or local storage by the RudderStack’s JavaScript SDK.
+---
+
+# Data Storage in Cookies
+
+RudderStack's JavaScript SDK stores persistent user data in cookies by default. In case the cookies are not supported, local storage is considered for the same. 
+
+All the cookies are stored in the top-level domain by default. It helps to identify the users visiting from websites that are hosted under a sub-domain. For example, if you include the RudderStack JavaScript SDK in both `admin.samplewebsite.com` and `app.samplewebsite.com`, the SDK will store the cookie in `samplewebsite.com`. However, you can configure where you want to store the cookie by using the `setCookieDomain` parameter in the `load` API options as shown:
```
Comment (ameypv-rudder) [inline stream-sources/rudderstack-sdk-integration-guides/rudderstack-javascript-sdk/data-storage-cookies.mdx:11]:
```suggestion
All the cookies are stored in the top-level domain by default. It helps you to identify the users visiting from websites hosted under a particular sub-domain. For example, if you include the JavaScript SDK in both `admin.samplewebsite.com` and `app.samplewebsite.com`, the SDK will store the cookie in `samplewebsite.com`. However, you can configure where you want to store the cookie by using the `setCookieDomain` parameter in the `load` API options as shown:
```

File: stream-sources/rudderstack-sdk-integration-guides/rudderstack-javascript-sdk/data-storage-cookies.mdx [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+
+```javascript
+rudderanalytics.load(<WRITE_KEY>, <DATA_PLANE_URL>, {
+  setCookieDomain: "samplewebsite.com",
+});
+```
+
+## Cookies
+
+The list of cookies used by JavaScript SDK to store persistent user data is mentioned below:
```
Comment (ameypv-rudder) [inline stream-sources/rudderstack-sdk-integration-guides/rudderstack-javascript-sdk/data-storage-cookies.mdx:21]:
```suggestion
The list of cookies used by the JavaScript SDK to store persistent user data is mentioned below:
```

File: stream-sources/rudderstack-sdk-integration-guides/rudderstack-javascript-sdk/data-storage-cookies.mdx [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+
+| Name             | Description       | Example    | Clearing Mechanism using SDK |
```
Comment (ameypv-rudder) [inline stream-sources/rudderstack-sdk-integration-guides/rudderstack-javascript-sdk/data-storage-cookies.mdx:23]:
```suggestion
| Name             | Description       | Example    | Clearing mechanism using the SDK |
```

File: stream-sources/rudderstack-sdk-integration-guides/rudderstack-javascript-sdk/data-storage-cookies.mdx [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+| :-------------------- | :---------------------------------- | :-------------------------------- |:---------------------------------- |
+|  `rl_user_id`         |Stores the user ID set via the `identify` API method. All the subsequent event payloads will contain this data unless cleared from the storage. | `4578`, `USER_001` | `rudderanalytics.reset()`  |
```
Comment (ameypv-rudder) [inline stream-sources/rudderstack-sdk-integration-guides/rudderstack-javascript-sdk/data-storage-cookies.mdx:25]:
```suggestion
|  `rl_user_id`         |Stores the user ID set via the `identify` API. All the subsequent event payloads will contain this data unless cleared from the storage. | `4578`, `USER_001` | `rudderanalytics.reset()`  |
```

File: stream-sources/rudderstack-sdk-integration-guides/rudderstack-javascript-sdk/data-storage-cookies.mdx [hunk 11]
```diff
@@ -0,0 +1,63 @@
...
+|  `rl_trait`       | Stores the user traits object set via the `identify` API method. All the subsequent event payloads will contain this data unless cleared from the storage. |  <pre>{<br>  email: "alex@example.com",<br>  accountType: "pro",<br>  country: "US",<br>  someObj: {<br>    key1: val1,<br>    key2: val2,<br>  }<br>}</pre>         |`rudderanalytics.reset()` |
```
Comment (ameypv-rudder) [inline stream-sources/rudderstack-sdk-integration-guides/rudderstack-javascript-sdk/data-storage-cookies.mdx:26]:
```suggestion
|  `rl_trait`       | Stores the user traits object set via the `identify` API. All the subsequent event payloads will contain this data unless cleared from the storage. |  <pre>{<br>  email: "alex@example.com",<br>  accountType: "pro",<br>  country: "US",<br>  someObj: {<br>    key1: val1,<br>    key2: val2,<br>  }<br>}</pre>         |`rudderanalytics.reset()` |
```

File: stream-sources/rudderstack-sdk-integration-guides/rudderstack-javascript-sdk/data-storage-cookies.mdx [hunk 12]
```diff
@@ -0,0 +1,63 @@
...
+|  `rl_group_id`          | Stores the user group ID set via the `group` API method. All the subsequent group event payloads will contain this data unless cleared from the storage.        | `GRP_3`, `98` | `rudderanalytics.reset()` |
+|  `rl_group_trait`    | Stores the user group traits object set via the `group` API method. All the subsequent group event payloads will contain this data unless cleared from the storage.   |    <pre>{<br>  location: "New Orleans",<br>  nationality: "US",<br>  someObj: {<br>    key1: val1,<br>    key2: val2,<br>  }<br>}</pre>                      |  `rudderanalytics.reset()` |
+| `rl_page_init_referrer`        | Stores the initial referrer of the page when a user visits a site for the first time. All the subsequent event payloads will contain this data.    | https://www.google.com/ | Cannot be cleared using SDK |
```
Comment (ameypv-rudder) [inline stream-sources/rudderstack-sdk-integration-guides/rudderstack-javascript-sdk/data-storage-cookies.mdx:30]:
```suggestion
| `rl_page_init_referrer`        | Stores the initial referrer of the page when a user visits a site for the first time. All the subsequent event payloads will contain this data.    | `https://www.google.com/` | Cannot be cleared using SDK |
```

File: stream-sources/rudderstack-sdk-integration-guides/rudderstack-javascript-sdk/data-storage-cookies.mdx [hunk 13]
```diff
@@ -0,0 +1,63 @@
...
+| `rl_page_init_referring_domain`      | Stores the initial referring domain of the page when a user visits a site for the first time. All the subsequent event payloads will contain this data.  |  google.com |Cannot be cleared using SDK |
```
Comment (ameypv-rudder) [inline stream-sources/rudderstack-sdk-integration-guides/rudderstack-javascript-sdk/data-storage-cookies.mdx:31]:
```suggestion
| `rl_page_init_referring_domain`      | Stores the initial referring domain of the page when a user visits a site for the first time. All the subsequent event payloads will contain this data.  |  `google.com` | Cannot be cleared using SDK. |
```

File: stream-sources/rudderstack-sdk-integration-guides/rudderstack-javascript-sdk/data-storage-cookies.mdx [hunk 14]
```diff
@@ -0,0 +1,63 @@
...
+
+<div class="infoBlock">
+
+The cookie values are encrypted and their length is directly proportional to the values provided to the SDK. Also, all the cookie names are prefixed with <code class="inline-code">rl_</code> and the values are prefixed with <code class="inline-code">RudderEncrypt:</code>. For example, <code class="inline-code">rl_user_id —> RudderEncrypt:U2FsdGVkX1+UKmiooYoGmKdNws7sgmWgGfHe</code>.
+</div>
+
+
+## Local storage cookies
+
+The local storage cookies are used by the JavaScript SDK to keep track of the events being sent to the RudderStack backend. These can be listed as below:
```
Comment (ameypv-rudder) [inline stream-sources/rudderstack-sdk-integration-guides/rudderstack-javascript-sdk/data-storage-cookies.mdx:41]:
```suggestion
The local storage cookies are used by the JavaScript SDK to keep track of the events sent to the RudderStack backend. They are listed in the table below:
```

File: stream-sources/rudderstack-sdk-integration-guides/rudderstack-javascript-sdk/data-storage-cookies.mdx [hunk 15]
```diff
@@ -0,0 +1,63 @@
...
+
+| Name             | Description       | Example    |
+| :------------- |:---------- | :----------- | 
+|  `ack`     | Timer for other browser tabs to claim control of the retry queue     | `1639734070124` |
```
Comment (ameypv-rudder) [inline stream-sources/rudderstack-sdk-integration-guides/rudderstack-javascript-sdk/data-storage-cookies.mdx:45]:
```suggestion
|  `ack`     | Timer for other browser tabs to claim control of the retry queue.     | `1639734070124` |
```

File: stream-sources/rudderstack-sdk-integration-guides/rudderstack-javascript-sdk/data-storage-cookies.mdx [hunk 16]
```diff
@@ -0,0 +1,63 @@
...
+|  `reclaimStart` and `reclaimEnd`      |Determines if a tab takes over the queue from another tab  | `2dc2aee6-2836-4273-be69-79c90c04ddec`  |
+|  `inProgress`    | Keeps track of the events in progress | <pre>{<br>  "d89d7fb5-945e-4378-bda5-492e4b596fb4": {<br>   "item": {<br>    "url": "https://rudderstack-dataplane.rudderstack.com/v1/track",<br>    "headers": {<br>     "Content-Type": "application/json",<br>      ...<br>    },<br>    "message": { <br>     ...<br>    }<br>    "attemptNumber": 1,<br>    "time": 1639734792773,<br>    "id": "a4d89d7f-b594-4eb3-b8bd-a5492e4b596f"<br>   }<br>  }<br>}<br></pre> |
```
Comment (ameypv-rudder) [inline stream-sources/rudderstack-sdk-integration-guides/rudderstack-javascript-sdk/data-storage-cookies.mdx:47]:
```suggestion
|  `inProgress`    | Keeps track of the events in progress. | <pre>{<br>  "d89d7fb5-945e-4378-bda5-492e4b596fb4": {<br>   "item": {<br>    "url": "https://rudderstack-dataplane.rudderstack.com/v1/track",<br>    "headers": {<br>     "Content-Type": "application/json",<br>      ...<br>    },<br>    "message": { <br>     ...<br>    }<br>    "attemptNumber": 1,<br>    "time": 1639734792773,<br>    "id": "a4d89d7f-b594-4eb3-b8bd-a5492e4b596f"<br>   }<br>  }<br>}<br></pre> |
```

File: stream-sources/rudderstack-sdk-integration-guides/rudderstack-javascript-sdk/data-storage-cookies.mdx [hunk 17]
```diff
@@ -0,0 +1,63 @@
...
+|  `queue`    | Keeps track of the events that are in queue to be processed   | <pre>[<br>  {<br>   "item": {<br>    "url": "https://rudderstack-dataplane.rudderstack.com/v1/track",<br>    "headers": {<br>     "Content-Type": "application/json",<br>      ...<br>    },<br>    "message": { <br>     ...<br>    }<br>    "attemptNumber": 0,<br>    "time": 1639734792773,<br>    "id": "a4d89d7f-b594-4eb3-b8bd-a5492e4b596f"<br>   }<br>  }<br>]<br></pre> |
```
Comment (ameypv-rudder) [inline stream-sources/rudderstack-sdk-integration-guides/rudderstack-javascript-sdk/data-storage-cookies.mdx:48]:
```suggestion
|  `queue`    | Keeps track of the events that are in queue to be processed.   | <pre>[<br>  {<br>   "item": {<br>    "url": "https://rudderstack-dataplane.rudderstack.com/v1/track",<br>    "headers": {<br>     "Content-Type": "application/json",<br>      ...<br>    },<br>    "message": { <br>     ...<br>    }<br>    "attemptNumber": 0,<br>    "time": 1639734792773,<br>    "id": "a4d89d7f-b594-4eb3-b8bd-a5492e4b596f"<br>   }<br>  }<br>]<br></pre> |
```

**Review by ameypv-rudder — APPROVED** [review]

**Pull request merged**